### PR TITLE
fix: warn the user if the JSON stats file already exists

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
fix

**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**
No
**Summary**
 warn the user if the JSON stats file already exists.

`file 'stats.json' already exists. CLI will overwrite the existing file.`

**Does this PR introduce a breaking change?**
No

**Other information**
No

### CODE REVIEW

File: packages/webpack-cli/lib/webpack-cli.js [hunk 1]
```diff
@@ -487,6 +487,9 @@ class WebpackCLI {
                 const JSONStats = JSON.stringify(stats.toJson(foundStats), null, 2);
 
                 try {
+                    if (existsSync(args.json)) {
+                        logger.info(`file '${args.json}' already exists. CLI will overwrite the existing file.`);
```
Comment (rishabh3112) [inline packages/webpack-cli/lib/webpack-cli.js:491]:
```suggestion
                        logger.info(`file '${args.json}' already exists and will be overwritten.`);
```

File: test/json/json.test.js [hunk 2]
```diff
@@ -76,6 +76,30 @@ describe('json flag', () => {
         });
     });
 
+    it('should warn the user if stats file already exist', (done) => {
+        let { stdout, exitCode } = run(__dirname, ['--json', 'stats.json']);
+        // run 2nd compilation
+        ({ stdout, exitCode } = run(__dirname, ['--json', 'stats.json']));
+
+        expect(stdout).toContain(`file 'stats.json' already exists. CLI will overwrite the existing file.`);
```
Comment (rishabh3112) [inline test/json/json.test.js:84]:
```suggestion
        expect(stdout).toContain(`file 'stats.json' already exists and will be overwritten.`);
```

File: test/json/json.test.js [hunk 3]
```diff
@@ -76,6 +76,30 @@ describe('json flag', () => {
         });
     });
 
+    it('should warn the user if stats file already exist', (done) => {
+        let { stdout, exitCode } = run(__dirname, ['--json', 'stats.json']);
```
Comment (rishabh3112) [inline test/json/json.test.js:80]:
```suggestion
       run(__dirname, ['--json', 'stats.json']);
```

**Review by rishabh3112 — CHANGES_REQUESTED** [review]
Left a few suggestions.
Maintain immutability.
Use simple sentences wherever possible.

File: test/json/json.test.js [hunk 4]
```diff
@@ -76,6 +76,30 @@ describe('json flag', () => {
...
+        // run 2nd compilation
+        ({ stdout, exitCode } = run(__dirname, ['--json', 'stats.json']));
```
Comment (rishabh3112) [inline test/json/json.test.js:82]:
```suggestion
        const { stdout, exitCode } = run(__dirname, ['--json', 'stats.json']);
```

**Review by rishabh3112 — CHANGES_REQUESTED** [review]

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:491]: Yeah It's better.

File: test/json/json.test.js [hunk 5]
```diff
@@ -76,6 +76,30 @@ describe('json flag', () => {
         });
     });
 
+    it('should warn the user if stats file already exist', (done) => {
+       run(__dirname, ['--json', 'stats.json']);
```
Comment (snitin315) [inline test/json/json.test.js:80]:
```suggestion
        run(__dirname, ['--json', 'stats.json']);
```

File: packages/webpack-cli/lib/webpack-cli.js [hunk 6]
```diff
@@ -487,6 +487,9 @@ class WebpackCLI {
                 const JSONStats = JSON.stringify(stats.toJson(foundStats), null, 2);
 
                 try {
+                    if (existsSync(args.json)) {
+                        logger.info(`file '${args.json}' already exists and will be overwritten.`);
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:491]: `logger. warn`

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:491]:
```suggestion
                        logger.warn(`file '${args.json}' already exists and will be overwritten.`);
```

File: test/json/json.test.js [hunk 7]
```diff
@@ -76,6 +76,30 @@ describe('json flag', () => {
         });
     });
 
+    it('should warn the user if stats file already exist', (done) => {
+        run(__dirname, ['--json', 'stats.json']);
```
Comment (alexander-akait) [inline test/json/json.test.js:80]: We should check `exitCode`/`stderr`/`stdout` here too

Comment (alexander-akait) [reply test/json/json.test.js:80]: As I said each tests should check `exitCode`/`stderr`/`stdout`

**Review by alexander-akait — COMMENTED** [review]
Should be merged after https://github.com/webpack/webpack-cli/pull/2144, approved

**Review by alexander-akait — COMMENTED** [review]
hm, what about if you in watch mode, it is expected behavior, and warning will be like a spam

**Comment by snitin315** [comment]
Just found `--watch` is not working as expected with `--json`. 
Screenshot at 2020-12-02 07-20-18

because we don't continue to print stats after passing `--json`

**Comment by alexander-akait** [comment]
Ideally it should work, we need `--interactive` for this

**Comment by snitin315** [comment]
Its working just stats are not being logged.

**Comment by alexander-akait** [comment]
We can add workaround for watch and --json, we just should remove all from screen before printing them again

**Comment by snitin315** [comment]
@alexander-akait can you clarify what should be the expected behavior?

**Comment by alexander-akait** [comment]
When you have file on first run we should log to user about file exists and it will be override, for other callbacks we should not log this + avoid using sync API

**Comment by snitin315** [comment]
I'm not sure how to achieve this behavior here - 

https://github.com/webpack/webpack-cli/blob/caa3348a825dfc5e0c6320304a28ce49630052ef/packages/webpack-cli/lib/webpack-cli.js#L1288-L1303

**Review by alexander-akait — COMMENTED** [review]
Pseudo core

```
if (isFirstRun && (await stats(args.json))) {
  logger.warn(`file '${args.json}' already exists and will be overwritten.`);
}

await writeFile(args.json, JSONStats);

isFirstRun = true;
```

**Pull request closed without merging**